### PR TITLE
[TC-460] Ds tenancy validation server

### PR DESCRIPTION
**Pull request by nir-sopher**

### CODE REVIEW

File: traffic_ops/app/lib/API/DeliveryServiceServer.pm [hunk 1]
```diff
@@ -27,6 +28,11 @@ sub index {
 	my @data;
 	my $orderby = $self->param('orderby') || "deliveryservice";
 
+	#FOR THE REVIEWER - Currently I do not check DS tenancy here.
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/DeliveryServiceServer.pm]: With the new roles/capabilities model there will be a lot of flexibility regarding roles and capabilities. For example, we could create a role called ds-admin with the following capabilities: 

- ds-read
- ds-write
- ds-server-read
- ds-server-write
- ds-regex-read
- etc

so that means anyone with this ds-admin role can manage all aspects of a ds, including changing the server assignments.

other companies may only want to attach ds-server-write to the a higher level role but we can't control that so my suggestion is we always check tenancy and filter results accordingly.

Comment (nir-sopher) [inline traffic_ops/app/lib/API/DeliveryServiceServer.pm]: As I saw it, the ds-admin should not be aware of specific servers existence, so this API should not be available to him.
However, as you see usecases a ds-admin may approach this API, I'll had the tenancy check

File: traffic_ops/app/lib/API/Server.pm [hunk 2]
```diff
@@ -400,40 +401,25 @@ sub get_servers_by_status {
 	my $orderby           = $self->param('orderby') || "hostName";
 	my $orderby_snakecase = lcfirst( decamelize($orderby) );
 
+	my $forbidden;
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: not sure why this method needed to change (get_servers_by_status) as it seems unrelated to delivery services.

File: traffic_ops/app/lib/API/Server.pm [hunk 3]
```diff
@@ -445,18 +431,27 @@ sub get_servers_by_dsid {
...
 	my @ds_servers;
 	my $forbidden;
-	if ( &is_privileged($self) || $self->is_delivery_service_assigned($ds_id) ) {
+	my $servers;
+
+	if (defined($ds) && !$tenant_utils->is_ds_resource_accessible($tenants_data, $ds->tenant_id)) {
+		$forbidden = "Forbidden. Delivery service not available for user's tenant.";
+		return ($forbidden, $servers);
+	}
+	elsif ( &is_privileged($self) || $tenant_utils->ignore_ds_users_table() || $self->is_delivery_service_assigned($ds_id) ) {
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: same thing. is this going to change to $tenant_utils->use_tenancy()

File: traffic_ops/app/lib/API/Server.pm [hunk 4]
```diff
@@ -489,11 +484,22 @@ sub get_edge_servers_by_dsid {
...
+	if ( !defined($ds) ) {
+		return $self->not_found();
+	}
+
 	my $ds_servers;
-	if ( &is_privileged($self) || $self->is_delivery_service_assigned($ds_id) ) {
+	my $tenant_utils = Utils::Tenant->new($self);
+	my $tenants_data = $tenant_utils->create_tenants_data_from_db();
+	if (!$tenant_utils->is_ds_resource_accessible($tenants_data, $ds->tenant_id)) {
+		return $self->forbidden();
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: can you add a forbidden message?

File: traffic_ops/app/lib/API/Server.pm [hunk 5]
```diff
@@ -569,10 +575,17 @@ sub get_unassigned_servers_by_dsid {
...
+	my $tenants_data = $tenant_utils->create_tenants_data_from_db();
+
+	if (!$tenant_utils->is_ds_resource_accessible($tenants_data, $ds->tenant_id)) {
+		return $self->forbidden();
+	}
+	elsif ( &is_privileged($self) || $tenant_utils->ignore_ds_users_table() || $self->is_delivery_service_assigned($ds_id) ) {
 		@assigned_servers = $self->db->resultset('DeliveryserviceServer')->search( \%ds_server_criteria, { prefetch => [ 'deliveryservice', 'server' ] } )->get_column('server')->all();
 	}
 	else {
+		#for the reviewer - I believe it should turn into forbidden as well
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: yes, this is wrong. should be $self->forbidden

File: traffic_ops/app/lib/API/Server.pm [hunk 6]
```diff
@@ -569,10 +575,17 @@ sub get_unassigned_servers_by_dsid {
 	$ds_server_criteria{'deliveryservice.id'} = $ds_id;
 
 	my @assigned_servers;
-	if ( &is_privileged($self) || $self->is_delivery_service_assigned($ds_id) ) {
+	my $tenant_utils = Utils::Tenant->new($self);
+	my $tenants_data = $tenant_utils->create_tenants_data_from_db();
+
+	if (!$tenant_utils->is_ds_resource_accessible($tenants_data, $ds->tenant_id)) {
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: you might want to rethink this if/elseif/else seems wrong.

File: traffic_ops/app/lib/API/Server.pm [hunk 7]
```diff
@@ -651,7 +664,14 @@ sub get_eligible_servers_by_dsid {
...
-	if ( !&is_privileged($self) && !$self->is_delivery_service_assigned($ds_id) ) {
+	my $tenant_utils = Utils::Tenant->new($self);
+	my $tenants_data = $tenant_utils->create_tenants_data_from_db();
+
+	if (!$tenant_utils->is_ds_resource_accessible($tenants_data, $ds->tenant_id)) {
+		return $self->forbidden();
+	}
+	elsif ( !&is_privileged($self) && !$tenant_utils->ignore_ds_users_table() && !$self->is_delivery_service_assigned($ds_id) ) {
+		#for the reviewer - I believe it should turn into forbidden as well
 		return $self->alert("Forbidden. Delivery service not assigned to user.");
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: yep, forbidden

File: traffic_ops/app/lib/API/Server.pm [hunk 8]
```diff
@@ -726,41 +746,26 @@ sub get_servers_by_type {
 	my $orderby           = $self->param('orderby') || "hostName";
 	my $orderby_snakecase = lcfirst( decamelize($orderby) );
 
+	my $forbidden;
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: not sure why this changed. seems unrelated to delivery services but if you are trying to establish consistency, i am fine with it.

File: traffic_ops/app/lib/API/Server.pm [hunk 9]
```diff
@@ -947,6 +952,10 @@ sub details_v11 {
 		}
 
 		my $rs_ds_data = $row->deliveryservice_servers;
+		#FOR THE REVIEWER - Currently I do not check DS tenancy here.
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: always check tenancy in my opinion if use_tenancy=1. never rely on roles / capabilities because roles can be setup to include whatever capabilities the system owner wants. there is no harm in always ensuring the user has the proper tenancy. they can always sidestep tenancy by assigning themself with the root tenant...

File: traffic_ops/app/lib/API/Server.pm [hunk 10]
```diff
@@ -1021,6 +1030,10 @@ sub details {
 			}
 
 			my $rs_ds_data = $row->deliveryservice_servers;
+			#FOR THE REVIEWER - Currently I do not check DS tenancy here.
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: same thing. always check tenancy for anything ds-related imo

File: traffic_ops/app/lib/API/DeliveryServiceServer.pm [hunk 11]
```diff
@@ -80,15 +86,21 @@ sub assign_servers_to_ds {
...
-
 	my $ds = $self->db->resultset('Deliveryservice')->find( { id => $ds_id } );
 	if ( !defined($ds) ) {
 		return $self->not_found();
 	}
 
+	my $tenant_utils = Utils::Tenant->new($self);
+	my $tenants_data = $tenant_utils->create_tenants_data_from_db();
+	if (!$tenant_utils->is_ds_resource_accessible($tenants_data, $ds->tenant_id)) {
+		return $self->alert("Invalid delivery-service assignment. The delivery-service is not avaialble for your tenant.");
```
Comment (nir-sopher) [inline traffic_ops/app/lib/API/DeliveryServiceServer.pm]: Are you sure?
This is not part of the path

$r->post("/api/$version/deliveryserviceserver")->over( authenticated => 1, not_ldap => 1 )->to( 'DeliveryServiceServer#assign_servers_to_ds', namespace => $namespace );

File: traffic_ops/app/lib/API/DeliveryServiceServer.pm [hunk 12]
```diff
@@ -223,8 +240,19 @@ sub remove_server_from_ds {
 	my $ds_id  	 	= $self->param('dsId');
 	my $server_id	= $self->param('serverId');
 
-	if ( !&is_privileged($self) && !$self->is_delivery_service_assigned($ds_id) ) {
-		$self->forbidden("Forbidden. Delivery service not assigned to user.");
+	my $tenant_utils = Utils::Tenant->new($self);
+	if ( !&is_privileged($self) && !$tenant_utils->ignore_ds_users_table() && !$self->is_delivery_service_assigned($ds_id) ) {
```
Comment (nir-sopher) [inline traffic_ops/app/lib/API/DeliveryServiceServer.pm]: Probably yes, we need to decide if we merge only the 2 parameters (use-tenancy & ignore_ds_user_assignment) or the functions as well.
If we merge also the fucntions, lets do it in one "search replace" commit

File: traffic_ops/app/lib/API/Server.pm (hunk 2 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: The function was previously testing the ds_user table as well and filtering by it.
We agreed it should be removed and replaced  only with "is-oper" test

Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: This function tested the "ds_user" table and filtered DSes accordingly.
We discussed it (by mail) and decided to remove this test and add only a "is-oper" test

File: traffic_ops/app/lib/API/Server.pm (hunk 3 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: Same, answer:)

File: traffic_ops/app/lib/API/Server.pm (hunk 4 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: Done

File: traffic_ops/app/lib/API/Server.pm (hunk 5 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: Fixed

File: traffic_ops/app/lib/API/Server.pm (hunk 7 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: done

File: traffic_ops/app/lib/API/Server.pm (hunk 8 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: This function previously filtered records by the ds_user table (only servers used for the DSes viewable to the user)
We agreed on mail to remove the test and put "is-oper"

File: traffic_ops/app/lib/API/Server.pm (hunk 6 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: Seems ok to me. 

case 1: "use=tenancy" == true, and resource is accessible to tenant:  
1. "if" test is false, continue
2. "elseif" test is true (as ignore_ds_users_table==true), so we get the list of servers

case 2: "use=tenancy" == true, and resource is not accessible to tenant:  
1. "if" test is true, return forbidden

case 3: "use=tenancy" == false, and ds is assigned to user
1. "if" test is false, continue
2. "elseif" test is true (as the ds is assigned to user), so we get the list of servers


case 3: "use=tenancy" == false, and ds is not assigned to user
1. "if" test is false, continue
2. "elseif" test is false (as the ds is not assigned to user),
3. We reach the failure

File: traffic_ops/app/lib/API/Server.pm (hunk 10 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: done

File: traffic_ops/app/lib/API/Server.pm (hunk 9 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Server.pm]: done

File: traffic_ops/app/lib/API/DeliveryServiceServer.pm (hunk 11 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/DeliveryServiceServer.pm]: ok, i see your point

File: traffic_ops/app/lib/API/DeliveryServiceServer.pm (hunk 12 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/DeliveryServiceServer.pm]: yep, i like the idea of maybe having 1 function use_tenancy() or whatever that keys off of the use-tenancy parameter

File: traffic_ops/app/lib/API/Server.pm (hunk 2 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: ok

File: traffic_ops/app/lib/API/Server.pm (hunk 6 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: ok, sounds good. thx for verifying

File: traffic_ops/app/lib/API/Server.pm (hunk 8 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Server.pm]: k
